Sources/OpenCombine/Publishers/ChildManagingInnerBase.swift

```diff
@@ -0,0 +1,327 @@
+        let shouldProcessQueue: Bool = lock.do {
+            lockedWork()
+            if childSubscribers.count == 0 && !lockedHasNonChildSubscription() {
+                sendFinishedAfterDrainingQueue = true
+                if !processingQueue {
+                    processingQueue = true
+                    return true
+                }
+            }
+            return false
+        }
+
+        if shouldProcessQueue {
+            processQueue()
+        }
+    }
+}
+
+// Private implementation
+extension InnerWithChildrenBase {
+    private func _lockedDeactivate() -> Set<ChildSubscriber> {
+        downstream = nil
+        downstreamDemand = .none
+        let result = childSubscribers
+        childSubscribers.removeAll()
+        lockedDeactivate()
+        return result
+    }
+
+    private func _unlockedDeactivate(_ childrenToCancel: Set<ChildSubscriber>) {
+        childrenToCancel.forEach { $0.cancel() }
+        unlockedDeactivate()
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       completedWith completion: Subscribers.Completion<Failure>) {
+        switch completion {
+        case .finished:
+            removeActiveSubscription(forChild: child)
+        case .failure:
+            downstream?.receive(completion: completion)
+            deactivate()
+        }
+    }
+
+    private func demandForChild() -> Subscribers.Demand {
+        return self.downstreamDemand == .unlimited ? .unlimited : .max(1)
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       receivedValue value: Output) -> Subscribers.Demand {
+        // Obtain the lock in order to determine what to do
+        let (valueWasBuffered, processTheQueue): (Bool, Bool) = lock.do {
+            // Consider pending values to send when assessing downstreamDemand
+            if downstreamDemand - valuesToSend.count == .none {
+                valuesToSend.append((value, child))
+                return (valueWasBuffered: true, drainTheQueue: false)
+            } else {
+                valuesToSend.append((value, nil))
+                if processingQueue {
+                    return (valueWasBuffered: false, drainTheQueue: false)
+                }
+                processingQueue = true
+                return (valueWasBuffered: false, drainTheQueue: true)
+            }
+        }
+
+        // TODO: Add test to make sure we demand .none when downstream is .unlimited
+        let demandResult: Subscribers.Demand = valueWasBuffered ? .none : .max(1)
+
+        if processTheQueue {
+            processQueue()
+        }
+
+        return demandResult
+    }
+
+    private func removeActiveSubscription(forChild child: ChildSubscriber) {
+        evaluateActiveSubscriptionsExecuting { childSubscribers.remove(child) }
+    }
+
+    private enum QueueWorkStatus {
+        case noWork
+        case sendFinish
+        case sendValues(values: [WorkType])
```

Comment:
Let's use `ArraySlice<WorkType>` here.
```suggestion
        case sendValues(values: ArraySlice<WorkType>)
```

This will allow us to avoid allocating a whole array below.